### PR TITLE
MIR borrowck: error message confuses locals and temporaries

### LINKED ISSUE #46471: MIR borrowck: error message confuses locals and temporaries
e.g.:
```Rust
fn foo() -> &u32 {
    let x = 0;
    &x
}
```

Gives the following error:
```
error[E0597]: `x` does not live long enough (Ast)
 --> foo.rs:4:6
  |
4 |     &x
  |      ^ does not live long enough
5 | }
  | - borrowed value only lives until here
  |
  = note: borrowed value must be valid for the static lifetime...

error[E0597]: borrowed value does not live long enough (Mir)
 --> foo.rs:5:2
  |
3 |     let x = 0;
  |         - temporary value created here
4 |     &x
5 | }
  | -^ temporary value dropped here while still borrowed
  | |
  | temporary value needs to live until here
  |
  = note: consider using a `let` binding to increase its lifetime

error: aborting due to 2 previous errors
```

The MIR error message should be more like the AST one

### LINKED ISSUE #46472: MIR borrowck: error does not know about free/named lifetimes
e.g.

```Rust
fn bar<'a>() -> &'a mut u32 {
    &mut 4
}
```

Gives the following errors for AST and MIR borrowck - the MIR error should be more like the AST error
```
error[E0597]: borrowed value does not live long enough (Ast)
 --> bar.rs:2:10
  |
2 |     &mut 4
  |          ^ does not live long enough
3 | }
  | - temporary value only lives until here
  |
note: borrowed value must be valid for the lifetime 'a as defined on the function body at 1:1...
 --> bar.rs:1:1
  |
1 | / fn bar<'a>() -> &'a mut u32 {
2 | |     &mut 4
3 | | }
  | |_^

error[E0597]: borrowed value does not live long enough (Mir)
 --> bar.rs:3:2
  |
2 |     &mut 4
  |          - temporary value created here
3 | }
  | -^ temporary value dropped here while still borrowed
  | |
  | temporary value needs to live until here
  |
  = note: consider using a `let` binding to increase its lifetime

error: aborting due to 2 previous errors
```

### PR DESCRIPTION
**Pull request by davidtwco**
Fixes #46471 and fixes #46472 (see [this Gitter comment](https://gitter.im/rust-impl-period/WG-compiler-nll?at=5a2d5cb53ae2aa6b3facf0c2)).

r? @arielb1

### CODE REVIEW

**Comment by davidtwco** [comment]
This isn't quite there yet, the error output is approximately correct, but the spans are a character to the left of what they should be to match exactly. I'm also not sure that the condition I'm checking is what I should be using, but it works.

**Comment by davidtwco** [comment]
I've fixed issues referenced in my previous comment and also fixed #46472 as mentioned in [this Gitter comment](https://gitter.im/rust-impl-period/WG-compiler-nll?at=5a2d5cb53ae2aa6b3facf0c2).

I think the function might be getting a bit a unwieldly since it's handling a bunch of cases. I've adjusted tests to allow the handful of places where spans are off by a character.

File: src/librustc_mir/borrow_check/mod.rs [hunk 1]
```diff
@@ -2373,24 +2371,56 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
...
-        err.note("consider using a `let` binding to increase its lifetime");
-
-        if let Some(end) = end_span {
-            err.span_label(end, "temporary value needs to live until here");
-        }
 
-        err.emit();
+        match &self.describe_place(place) {
+            Some(description) => {
+                let mut err = self.tcx.path_does_not_live_long_enough(
```
Comment (arielb1) [inline src/librustc_mir/borrow_check/mod.rs]: This doesn't handle the case where a local needs to live until an `end_span`, aka
```Rust
let x = {
    let y = 0;
    &y
};
```

**Comment by arielb1** [comment]
Nice!

@bors r+ p=1

This should go before #46558, which changes the same code.

**Comment by arielb1** [comment]
@bors p=0

This hasn't passed travis yet

**Comment by kennytm** [comment]
@bors r-

Travis failed.

```
[00:57:14] failures:
[00:57:14]     [ui] ui/nll/capture-ref-in-struct.rs
[00:57:14]     [ui] ui/nll/closure-requirements/escape-argument.rs
[00:57:14]     [ui] ui/nll/closure-requirements/escape-upvar-nested.rs
[00:57:14]     [ui] ui/nll/closure-requirements/escape-upvar-ref.rs
```

File: src/librustc_mir/borrow_check/mod.rs (hunk 1 above)
Comment (davidtwco) [reply src/librustc_mir/borrow_check/mod.rs]: This should now be being handled.

**Comment by davidtwco** [comment]
I've fixed the previous issue where a local might need to live until an `end_span`. I've not looked at the other tests which are failing in the Travis build - do these just need updated to have the improved error messages or is this an issue with this PR @arielb1?

**Comment by davidtwco** [comment]
I've updated the existing tests to use the new error messages. This should now pass Travis.

File: src/librustc_mir/borrow_check/mod.rs [hunk 2]
```diff
@@ -2373,23 +2371,71 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
...
+            None => (false, "borrowed value".to_owned()),
+        };
+
+        let err_span = match borrow.region {
+            // See test ui/issue-46471-1.rs for this case.
+            RegionKind::ReScope(_) => span,
+            // See test ui/issue-46471.rs for this case.
+            RegionKind::ReEarlyBound(_) | RegionKind::ReFree(_) => proper_span,
+            // See test compile-fail/issue-36082.rs for this case.
+            _ if name == "borrowed value" => span,
```
Comment (arielb1) [inline src/librustc_mir/borrow_check/mod.rs]: nit: use  `if !with_name`

File: src/librustc_mir/borrow_check/mod.rs [hunk 3]
```diff
@@ -2373,23 +2371,71 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
...
+            // See test ui/issue-46472.rs for this case.
+            _ => borrow_span,
+        };
+
+        let mut err = self.tcx.path_does_not_live_long_enough(err_span, &name, Origin::Mir);
+        match borrow.region {
```
Comment (arielb1) [inline src/librustc_mir/borrow_check/mod.rs]: There looks to be quite some connection between the span and the error message, so I would prefer to have 4 separate functions - 1 for each distinct error messagee

**Comment by kennytm** [comment]
<details><summary>Travis is still failing.</summary>

```
[00:59:30] ---- [compile-fail] compile-fail/borrowck/borrowck-local-borrow-outlives-fn.rs stdout ----
[00:59:30] 	
[00:59:30] error in revision `mir`: /checkout/src/test/compile-fail/borrowck/borrowck-local-borrow-outlives-fn.rs:15: unexpected error: '15:5: 15:7: `x` does not live long enough [E0597]'
[00:59:30] 
[00:59:30] error in revision `mir`: /checkout/src/test/compile-fail/borrowck/borrowck-local-borrow-outlives-fn.rs:16: expected error not found: borrowed value does not live long enough
[00:59:30] 
[00:59:30] error in revision `mir`: 1 unexpected errors found, 1 expected errors not found
...
[00:59:30] ---- [compile-fail] compile-fail/borrowck/borrowck-local-borrow-with-panic-outlives-fn.rs stdout ----
[00:59:30] 	
[00:59:30] error in revision `mir`: /checkout/src/test/compile-fail/borrowck/borrowck-local-borrow-with-panic-outlives-fn.rs:16: unexpected error: '16:15: 16:23: `z.1` does not live long enough [E0597]'
[00:59:30] 
[00:59:30] error in revision `mir`: /checkout/src/test/compile-fail/borrowck/borrowck-local-borrow-with-panic-outlives-fn.rs:18: expected error not found: [E0597]
[00:59:30] 
[00:59:30] error in revision `mir`: 1 unexpected errors found, 1 expected errors not found
...
[00:59:30] ---- [compile-fail] compile-fail/region-borrow-params-issue-29793-big.rs stdout ----
[00:59:30] 	
[00:59:30] error in revision `mir`: /checkout/src/test/compile-fail/region-borrow-params-issue-29793-big.rs:84: unexpected error: '84:5: 84:6: `x` does not live long enough [E0597]'
[00:59:30] 
[00:59:30] error in revision `mir`: /checkout/src/test/compile-fail/region-borrow-params-issue-29793-big.rs:84: unexpected error: '84:5: 84:6: `y` does not live long enough [E0597]'
[00:59:30] 
[00:59:30] error in revision `mir`: /checkout/src/test/compile-fail/region-borrow-params-issue-29793-big.rs:84: expected error not found: borrowed value does not live long enough
[00:59:30] 
[00:59:30] error in revision `mir`: /checkout/src/test/compile-fail/region-borrow-params-issue-29793-big.rs:84: expected error not found: borrowed value does not live long enough
[00:59:30] 
[00:59:30] error in revision `mir`: 2 unexpected errors found, 2 expected errors not found
```

</details>

If nothing goes wrong, #46558 will be merged soon, so you'll likely need to rebase.

**Comment by arielb1** [comment]
@kennytm 

We're working on it.

**Comment by davidtwco** [comment]
All previous issues should be resolved and it's now rebased on top of the current master (or as of an hour ago).

**Comment by davidtwco** [comment]
Alright, fifth time lucky. I've fixed the indentation and other requested changes from Gitter and updated the failing tests from the recent Travis run.

**Comment by arielb1** [comment]
That's cool!

@bors r+

**Pull request merged**